### PR TITLE
Calo-L1 unpacker modified to read 5BX payload format

### PR DESCRIPTION
**Pull request by hftsoi**
#### PR description:
This PR is about the trigger Calo-L1 unpacker modified to read ECAL's 5BX FAT events properly. The current central unpacker for FAT events is assuming a payload size of 192 * 32-bit words and the same structure as for normal events, repeated itself one after another five times for each FAT event, and each Calo fragment contains one BX only. However this is not compatible with the actual current payload format which has a larger size of 5 * 192 * 32-bit words, where each Calo fragment contains all 5 BX already, stacked in the order BX-2, -1, 0, +1, +2 inside.

The modification involves duplication of all necessary unpacker functions from normal events unpacking, then to make changes there for FAT events unpacking. The idea is to have two separate unpackers, one for unpacking normal events only, and another for unpacking FAT events only, in order to minimize risk.

Note: UCTCTP7RawData5BX.h is duplicated from UCTCTP7RawData.h, which contains the index equations for locating Calo fragments in payload data, offsets are modified for 5BX, while geometry, word structure and etc. remain the same.

#### PR validation:
Validated by running private online DQM on recent CRUZET runs, all DQM plots looked expected with modified unpacker. Also asked the central DQM team to include these changes into temporary production for this last week of running, all tests ran fine and DQM plots looked expected.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc [hunk 1]
```diff
@@ -227,6 +224,195 @@ namespace l1t {
...
+          cPhi = 71;
+        else if (cPhi < -1) {
+          LogError("CaloLayer1Unpacker") << "Major error in makeECalTPGs5BX" << std::endl;
+          return;
+        }
+        for (int cEta = -28; cEta <= 28; cEta++) {  // Calorimeter Eta indices (HB/HE for now)
+          if (cEta != 0) {                          // Calorimeter eta = 0 is invalid
+            bool negativeEta = false;
+            if (cEta < 0)
+              negativeEta = true;
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc:251]:
```suggestion
            bool negativeEta = (cEta < 0);
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc [hunk 2]
```diff
@@ -227,6 +224,195 @@ namespace l1t {
...
+          cPhi = 71;
+        else if (cPhi < -1) {
+          LogError("CaloLayer1Unpacker") << "Major error in makeHCalTPGs5BX" << std::endl;
+          return;
+        }
+        for (int cEta = -28; cEta <= 28; cEta++) {  // Calorimeter Eta indices (HB/HE for now)
+          if (cEta != 0) {                          // Calorimeter eta = 0 is invalid
+            bool negativeEta = false;
+            if (cEta < 0)
+              negativeEta = true;
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc:305]:
```suggestion
            bool negativeEta = (cEta < 0);
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc [hunk 3]
```diff
@@ -227,6 +224,195 @@ namespace l1t {
...
+
+    void CaloLayer1Unpacker::makeHFTPGs5BX(uint32_t lPhi,
+                                           UCTCTP7RawData5BX& ctp7Data5BX,
+                                           HcalTrigPrimDigiCollection* hcalTPGs,
+                                           uint32_t BX_n) {
+      UCTCTP7RawData5BX::CaloType cType = UCTCTP7RawData5BX::HF;
+      for (uint32_t side = 0; side <= 1; side++) {
+        bool negativeEta = false;
+        if (side == 0)
+          negativeEta = true;
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc:350]:
```suggestion
        bool negativeEta = (cEta < 0);
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc [hunk 4]
```diff
@@ -227,6 +224,195 @@ namespace l1t {
...
+    }
+
+    void CaloLayer1Unpacker::makeRegions5BX(uint32_t lPhi,
+                                            UCTCTP7RawData5BX& ctp7Data5BX,
+                                            L1CaloRegionCollection* regions,
+                                            uint32_t BX_n) {
+      for (uint32_t side = 0; side <= 1; side++) {
+        bool negativeEta = false;
+        if (side == 0)
+          negativeEta = true;
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc:405]:
```suggestion
        bool negativeEta = (cEta < 0);
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc [hunk 5]
```diff
@@ -227,6 +224,195 @@ namespace l1t {
...
+        for (uint32_t region = 0; region <= 6; region++) {
+          uint32_t regionData = ctp7Data5BX.getRegionSummary(negativeEta, region, BX_n);
+          uint32_t lEta = 10 - region;  // GCT eta goes 0-21, 0-3 -HF, 4-10 -B/E, 11-17 +B/E, 18-21 +HF
+          if (!negativeEta)
+            lEta = region + 11;
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Unpacker.cc:410]:
```suggestion
          // GCT eta goes 0-21, 0-3 -HF, 4-10 -B/E, 11-17 +B/E, 18-21 +HF
          uint32_t lEta = negativeEta? 10 - region: region + 11;
```

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/UCTCTP7RawData5BX.h [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  bool isTowerMasked(CaloType cType, bool negativeEta, uint32_t cEta, uint32_t iPhi, uint32_t BX_n) {
+    uint32_t linkStatus = getLinkStatus(cType, negativeEta, cEta, iPhi, BX_n);
+    uint32_t tower = iPhi;
+    if ((cEta % 2) == 0)
+      tower += 4;
+    if (cType == HF) {
+      tower = (cEta - 30);
+      if (cEta == 41)
+        tower = 10;
+    }
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/UCTCTP7RawData5BX.h:311]:
```suggestion
    uint32_t tower;
    if (cType == HF) {
      tower = (cEta != 41? cEta - 30: 10);
    }  else {
      tower = iPhi;
      if ((cEta % 2) == 0)
        tower += 4; 
    }
```

**Review by perrotta — COMMENTED** [review]
Some (minor) comment for some (very minor) code optimization

**Comment by hftsoi** [comment]
> Some (minor) comment for some (very minor) code optimization

Thanks for your review and suggestion! I guess at this stage we hope to keep the duplicated parts (for 5BX unpacking only) with changes as minimal as possible compared to the original (for normal unpacking only), even if it is for code optimization, just to have a more consistent comparison in case things arise. We will consider it later for both duplicated and original parts the same time.

**Comment by cecilecaillol** [comment]
+l1

**Comment by cecilecaillol** [comment]
> 

@hftsoi Ok, please keep in mind the comments when you change both the original and duplicated parts

**Comment by perrotta** [comment]
+1
- I understand that code cleaning will be dealt with in a separate PR

**Pull request merged**